xbmc/utils/GUIUtils.cpp

```diff
@@ -0,0 +1,363 @@
+}
+
+void CGUIUtils::PlaySong(CFileItem& item)
+{
+  if (item.HasMusicInfoTag())
+    CServiceBroker::GetPlaylistPlayer().Play(std::make_shared<CFileItem>(item), "");
+}
+
+void CGUIUtils::AddToPlayList(const CFileItemPtr& pItem, CFileItemList& queuedItems)
+{
+  if (!pItem->CanQueue() || pItem->IsRAR() || pItem->IsZIP() || pItem->IsParentFolder())
+    return;
+
+  // fast lookup is needed here
+  queuedItems.SetFastLookup(true);
+
+  if (pItem->IsMusicDb() && pItem->m_bIsFolder && !pItem->IsParentFolder())
+  { // we have a music database folder, just grab the "all" item underneath it
+    XFILE::CMusicDatabaseDirectory dir;
+
+    if (!dir.ContainsSongs(pItem->GetPath()))
+    { // grab the ALL item in this category
+      // Genres will still require 2 lookups, and queuing the entire Genre folder
+      // will require 3 lookups (genre, artist, album)
+      CMusicDbUrl musicUrl;
+      if (musicUrl.FromString(pItem->GetPath()))
+      {
+        musicUrl.AppendPath("-1/");
+        CFileItemPtr item(new CFileItem(musicUrl.ToString(), true));
+        item->SetCanQueue(true); // workaround for CanQueue() check above
+        AddToPlayList(item, queuedItems);
+      }
+      return;
+    }
+  }
+  if (pItem->m_bIsFolder)
+  {
+    // Check if we add a locked share
+    if (pItem->m_bIsShareOrDrive)
+    {
+      CFileItem item = *pItem;
+      if (!g_passwordManager.IsItemUnlocked(&item, "music"))
+        return;
+    }
+
+    // recursive
+    CFileItemList items;
+    GetDirectory(pItem->GetPath(), items);
+    FormatAndSort(items);
+    for (int i = 0; i < items.Size(); ++i)
+      AddToPlayList(items[i], queuedItems);
+  }
+  else
+  {
+    if (pItem->IsPlayList())
+    {
+      std::unique_ptr<PLAYLIST::CPlayList> pPlayList(PLAYLIST::CPlayListFactory::Create(*pItem));
+      if (pPlayList)
+      {
+        // load it
+        if (!pPlayList->Load(pItem->GetPath()))
+        {
+          HELPERS::ShowOKDialogText(CVariant{6}, CVariant{477});
+          return;
+        }
+
+        PLAYLIST::CPlayList playlist = *pPlayList;
+        for (int i = 0; i < playlist.size(); ++i)
+        {
+          AddToPlayList(playlist[i], queuedItems);
+        }
+        return;
+      }
+    }
+    else if (pItem->IsInternetStream() && !pItem->IsMusicDb())
+    { // just queue the internet stream, it will be expanded on play
+      queuedItems.Add(pItem);
+    }
+    else if (pItem->IsPlugin() && pItem->GetProperty("isplayable").asBoolean())
+    {
+      // python files can be played
+      queuedItems.Add(pItem);
+    }
+    else if (!pItem->IsNFO() && (pItem->IsAudio() || pItem->IsVideo()))
+    {
+      CFileItemPtr itemCheck = queuedItems.Get(pItem->GetPath());
+      if (!itemCheck || itemCheck->m_lStartOffset != pItem->m_lStartOffset)
+      { // add item
+        CFileItemPtr item(new CFileItem(*pItem));
+        if (!m_musicDatabase.IsOpen())
+          m_musicDatabase.Open();
+        m_musicDatabase.SetPropertiesForFileItem(*item);
```

Comment:
Why is the db a member var? Couldn't it just be local variable and created, opened, used and **closed** here? This would be a piece of code I understand. To open of the db as a member here and close it somewhere else I cannot see how it is related seems awkward. To work without resource leaks, this implies a calling sequence that is not actually enforced, right?

Reply:
Its a member var and closed elsewhere because this part of the code is called recursively on every item being added to the playlist queue. Otherwise, the music db would be opened and closed for each item added, which could be potentially thousands.  It therefore makes sense to me to open it once, re-use that and then close it when finished. TBH, I have no idea if opening it each time would cause much of a performance hit but my gut feeling is that it would.